docs/codeql/codeql-language-guides/analyzing-data-flow-in-ruby.rst

```diff
@@ -0,0 +1,387 @@
+Due to the control-flow graph being split, there can be multiple data-flow and control-flow nodes associated with a single expression AST node.
+
+The predicate ``localFlowStep(Node nodeFrom, Node nodeTo)`` holds if there is an immediate data flow edge from the node ``nodeFrom`` to the node ``nodeTo``.
+You can apply the predicate recursively, by using the ``+`` and ``*`` operators, or you can use the predefined recursive predicate ``localFlow``.
+
+For example, you can find flow from an expression ``source`` to an expression ``sink`` in zero or more local steps:
+
+.. code-block:: ql
+
+     DataFlow::localFlow(source, sink)
+
+Using local taint tracking
+~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Local taint tracking extends local data flow by including non-value-preserving flow steps.
+For example:
+
+.. code-block:: ruby
+
+     temp = x
+     y = temp + ", " + temp
+
+If ``x`` is a tainted string then ``y`` is also tainted.
+
+The local taint tracking library is in the module ``TaintTracking``.
+Like local data flow, a predicate ``localTaintStep(DataFlow::Node nodeFrom, DataFlow::Node nodeTo)`` holds if there is an immediate taint propagation edge from the node ``nodeFrom`` to the node ``nodeTo``.
+You can apply the predicate recursively, by using the ``+`` and ``*`` operators, or you can use the predefined recursive predicate ``localTaint``.
+
+For example, you can find taint propagation from an expression ``source`` to an expression ``sink`` in zero or more local steps:
+
+.. code-block:: ql
+
+     TaintTracking::localTaint(source, sink)
+
+
+Using local sources
+~~~~~~~~~~~~~~~~~~~
+
+When asking for local data flow or taint propagation between two expressions as above, you would normally constrain the expressions to be relevant to a certain investigation.
+The next section will give some concrete examples, but there is a more abstract concept that we should call out explicitly, namely that of a local source.
+
+A local source is a data-flow node with no local data flow into it.
+As such, it is a local origin of data flow, a place where a new value is created.
+This includes parameters (which only receive global data flow) and most expressions (because they are not value-preserving).
+The class ``LocalSourceNode`` represents data-flow nodes that are also local sources.
+It comes with a useful member predicate ``flowsTo(DataFlow::Node node)``, which holds if there is local data flow from the local source to ``node``.
+
+Examples
+~~~~~~~~
+
+This query finds the filename argument passed in each call to ``File.open``:
+
+.. code-block:: ql
+
+    import codeql.ruby.DataFlow
+    import codeql.ruby.ApiGraphs
+    
+    from DataFlow::CallNode call
+    where call = API::getTopLevelMember("File").getAMethodCall("open")
+    select call.getArgument(0)
+
+Notice the use of the ``API`` module for referring to library methods.
+For more information, see ":doc:`Using API graphs in Ruby <using-api-graphs-in-ruby>`."
```

Comment:
This is causing the most recently reported Sphinx failure. The article does not exist on this branch, it's in Arthur's PR: https://github.com/github/codeql/pull/10957.

Reply:
@nickrolfe - I'm not sure of the best approach for this link. If we expect Arthur's PR to be merged soon, then we can just wait and update from `main` to get the tests passing and to be able to review the generated article. Otherwise, we might need to comment it out temporarily.